static/js/controllers/contacts-edit.js

```diff
@@ -192,142 +193,121 @@ angular.module('inboxControllers').controller('ContactsEditCtrl',
+      _.each(schema.fields, function(fieldConf, fieldName) {
+        var customType = fieldConf.type.match(/^(db|custom):(.*)$/);
         if (customType) {
-          result = result.then(function(doc) {
-            if(!doc[f]) {
-              doc[f] = {};
-            } else if(doc[f] === 'NEW') {
-              var extra = extras[f];
-              extra.type = customType[2];
-              extra.reported_date = Date.now();
-
-              var isChild = extra.parent === 'PARENT' ||
-                            (!extra.parent && conf.parent === 'PARENT');
-              if (isChild) {
-                delete extra.parent;
-              }
-
-              return persist(extra)
-                .then(function(newlySavedDoc) {
-                  doc[f] = newlySavedDoc;
-                  if(isChild) {
-                    children.push(newlySavedDoc);
-                  }
-                  return doc;
-                });
-            } else if(original && original[f] && doc[f] === original[f]._id) {
-              doc[f] = original[f];
+          if(!doc[fieldName]) {
+            doc[fieldName] = {};
+          } else if(doc[fieldName] === 'NEW') {
+            var preparedSibling = prepare(siblings[fieldName]);
+            preparedSibling.type = customType[2];
+
+
+            var isChild = preparedSibling.parent === 'PARENT' ||
+                          (!preparedSibling.parent && fieldConf.parent === 'PARENT');
+            if (isChild) {
+              delete preparedSibling.parent;
+              // Cloning to avoid the circular reference we would make:
+              //   doc.fieldName.parent.fieldName.parent...
+              doc[fieldName] = _.clone(preparedSibling);
+              preparedSibling.parent = doc;
             } else {
-              var docId = doc[f];
-              if (typeof docId === 'object') {
-                docId = doc[f]._id;
-              }
-              return DB().get(doc[f])
-                .then(function(dbFieldValue) {
-                  doc[f] = dbFieldValue;
-                  return doc;
-                });
+              doc[fieldName] = preparedSibling;
             }
-            return doc;
-          });
+
+            preparedSiblings.push(preparedSibling);
+          } else if(original &&
+                    original[fieldName] &&
+                    original[fieldName]._id === doc[fieldName]) {
+
+            doc[fieldName] = original[fieldName];
+          } else {
+            throw new Error('Unsupported operation, talk to your friendly dev');
+            // REVIEWER:
+            //
+            // I don't think this branch is actually exercised, when:
+            //   var doc = doc[fieldName];
+            //  either:
+            //   if (typeof docId === 'object')
+            // is misleading and never exercised, or there is a bug in:
+            //   DB().get(doc[fieldName])
+            // since doc[fieldName] it could be an object not a string _id
```

Comment:
I think it's used and I'm not sure there is a bug. If `doc[fieldName]` is an object it gets `doc[fieldName]._id` which is almost certainly a string. It's really hard to read but that's my interpretation.